### PR TITLE
Feature user permissions

### PR DESCRIPTION
**Pull request by EugenioCorso**
Module for the creation of "staff" permission and his grant to other user from staff members

### CODE REVIEW

**Comment by libremente** [comment]
@EugenioCorso @CorsoGabriele Code looks fine to me. Do we have some tests?

File: fare/records/config.py [hunk 1]
```diff
@@ -46,10 +47,10 @@ def _(x):
         default_media_type='application/json',
         max_result_window=10000,
         error_handlers=dict(),
-        create_permission_factory_imp=allow_all,
+        create_permission_factory_imp=deny_all,
```
Comment (libremente) [inline fare/records/config.py]: Could you please explain me why we need a `deny_all` rule? Isn't this clashing with the permission system in place? Does this mean that if I create a RESTful client and try to connect to this server I will get a 500? I would like to have a functioning RESTful server so this `deny_all` is kinda scary (but maybe there is a logic behind it!). Thanks!

Comment (EugenioCorso) [reply fare/records/config.py]: With the `allow_all` permission every users can create a record composed only with its metadata and no content attached to it, that is not what we want. In order to do so @CorsoGabriele and I have created the view in the file_management module, in this way we can also check if the user has the correct permission to create a record. Now we are working to create the test for this part

Comment (libremente) [reply fare/records/config.py]: Ok, great

**Comment by libremente** [comment]
Needs to be rebased against new master

**Comment by libremente** [comment]
@EugenioCorso @CorsoGabriele 
Automated tests cannot even run due to this:
```
fare/grant_staff/api.py:1 at module level:

        D100: Missing docstring in public module

fare/grant_staff/api.py:10 in public function `grant_staff_permission`:

        D205: 1 blank line required between summary line and description (found 0)

fare/grant_staff/api.py:10 in public function `grant_staff_permission`:

        D208: Docstring is over-indented

fare/grant_staff/api.py:10 in public function `grant_staff_permission`:

        D210: No whitespaces allowed surrounding docstring text

fare/grant_staff/api.py:10 in public function `grant_staff_permission`:

        D400: First line should end with a period (not 'r')

tests/conftest.py:31 in public function `app`:

        D202: No blank lines allowed after function docstring (found 1)

tests/conftest.py:78 in public function `users`:

        D202: No blank lines allowed after function docstring (found 1)
```

See https://travis-ci.org/open-education-polito/fare-invenio/jobs/650879485?utm_medium=notification 
Could you please fix and push? I am interested to see the tests results in order to proceed with the debugging. Thanks!

**Comment by EugenioCorso** [comment]
@libremente i have fixed the issues, now the automated tests should work. Notice that 'test_simple_flow' in the test_api_simple_flow.py will always fail because of the permission's modification we made in fare/records/config.py so it should be removed

File: fare/grant_staff/api.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    user, role = _datastore._prepare_role_modify_args(user, role)
+
+    # check if the user exist
+    if user is None:
+        return 404
+    # if successfull print it in the terminal with green color
+    if _datastore.add_role_to_user(user, role):
+        print('\x1b[0;32;40m' + 'Role "' + str(role) + '" added '
+              'to user "' + str(user) + '" successfully.' + '\x1b[0m')
```
Comment (libremente) [inline fare/grant_staff/api.py]: Where will this be printed?

File: tests/test_user_permission.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+def test_anonymous_user_get_page(app):
+    """
+    Test purpose:
+    an anonymous user perform a GET to the view for granting permissions
+
+    Input:
+    non authenticated user
+
+    Output:
+    GET of the permission page to change permission with status code 301
```
Comment (libremente) [inline tests/test_user_permission.py]: 301?

File: tests/test_user_permission.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    Output:
+    GET return grant permission page with status code 200
+    """
+
+    url = "https://localhost:5000/grant_staff/grant"
+    staff = users['staff']
+    login_user(staff)
+    r = requests.get(url, verify=False)
+    assert r.status_code == 200
```
Comment (libremente) [inline tests/test_user_permission.py]: I don't think this is exactly what we want to test. 
This is the HTTP code returned which is OK, however I believe this is just *part* of the test: the most important one should be to check whether the staff role has been granted or not. 
I'd like to see something like:
```
assert user !staff
perform actions
assert user staff or still !staff (depending on the actions performed)
```
What do you think? We have direct DB access so we can query it and see what the actions really do

**Comment by libremente** [comment]
Also, @EugenioCorso please rebase against new master, thanks!

**Comment by libremente** [comment]
> @libremente i have fixed the issues, now the automated tests should work. Notice that 'test_simple_flow' in the test_api_simple_flow.py will always fail because of the permission's modification we made in fare/records/config.py so it should be removed

@EugenioCorso if it should be removed then please go on in this PR and remove it, thanks

File: fare/grant_staff/api.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    param string user: the email of the user that obtain staff permission
+    """
+    role = "staff"
+
+    user, role = _datastore._prepare_role_modify_args(user, role)
+
+    # check if the user exist
+    if user is None:
+        return 404
```
Comment (libremente) [inline fare/grant_staff/api.py]: Everytime we hit an error it is mandatory to log it. I think we have to define a logging system and use it all over the functions implemented, especially in cases like this.

Comment (libremente) [reply fare/grant_staff/api.py]: See #51

File: fare/grant_staff/api.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from flask import current_app
+from werkzeug.local import LocalProxy
+
+_datastore = LocalProxy(lambda: current_app.extensions['security'].datastore)
+
+
+def grant_staff_permission(user):
+    """Grant staff permission to  a user.
+
+    param string user: the email of the user that obtain staff permission
```
[the remark this answers is not in the document]
Comment (EugenioCorso) [reply fare/grant_staff/api.py]: I've renamed it to user_email, in this way it's clear. Thanks

**Comment by libremente** [comment]
Please rebase and mark the PR as ready for review @EugenioCorso

**Pull request merged**